### PR TITLE
Two pidfd fixes

### PR DESCRIPTION
**Pull request by daandemeyer**

### CODE REVIEW

File: src/basic/cgroup-util.c [hunk 1]
```diff
@@ -149,7 +149,9 @@ int cg_read_pidref(FILE *f, PidRef *ret, CGroupFlags flags) {
                 if (pid == 0)
                         return -EREMOTE;
 
-                if (FLAGS_SET(flags, CGROUP_NO_PIDFD)) {
+                /* We might read kernel thread pids from cgroup.procs for which we cannot create a pidfd so
+                 * catch those and don't try to create a pidfd for them. */
+                if (FLAGS_SET(flags, CGROUP_NO_PIDFD) || pid_is_kernel_thread(pid)) {
```
Comment (YHNdnzj) [inline src/basic/cgroup-util.c]:
```suggestion
                if (FLAGS_SET(flags, CGROUP_NO_PIDFD) || pid_is_kernel_thread(pid) > 0) {
```

**Review by bluca — APPROVED** [review]

File: src/basic/cgroup-util.c [hunk 2]
```diff
@@ -149,7 +149,9 @@ int cg_read_pidref(FILE *f, PidRef *ret, CGroupFlags flags) {
                 if (pid == 0)
                         return -EREMOTE;
 
-                if (FLAGS_SET(flags, CGROUP_NO_PIDFD)) {
+                /* We might read kernel thread pids from cgroup.procs for which we cannot create a pidfd so
+                 * catch those and don't try to create a pidfd for them. */
+                if (FLAGS_SET(flags, CGROUP_NO_PIDFD) || pid_is_kernel_thread(pid) > 0) {
```
Comment (poettering) [inline src/basic/cgroup-util.c:154]: uh, we cannot create pidfds of kernel threads? that's a problem, no? context?

**Comment by poettering** [comment]
I am a bit puzzled about this. the 2nd patch seems unnecessary? I had no trouble creating pidfds of kernel threads when i hacked this up? why would the kernel not allow that? what am missing?

**Comment by DaanDeMeyer** [comment]
@poettering It's a little messy. You could create pidfds of kernel threads, but when you send signals to them you got `EPERM` when not running as root, which caused cgroup kill failures. I chatted with @brauner who landed a patch in the kernel to disallow creating pidfds for kernel threads, as he didn't see a proper use case for it, which is why I wrote the comment like that.

However this ended up breaking other use cases, so that patch is getting reverted. So we should probably update the comment. Still, I think @brauner would prefer we don't create pidfds for kernel threads as pidfds were not envisioned to be used for kernel threads (yet).

**Pull request merged**